### ISSUE TITLE
Current location does not always update properly

### ISSUE THREAD

**Issue by jmcmahon356** [FasterXML/jackson-dataformat-smile#24]
While running through a data file containing a Smile-formatted array of objects, it was noticed that `getCurrentLocation().getByteOffset()` in SmileParser.java occasionally went backwards compared to the previous value after a read.  The parser was initialized with an `InputStream` that pointed to the aforementioned file.

Tracing through the code, the problem appears to occur in the `_loadToHaveAtLeast(int minAvailable)` method.  While deserializing one of the objects in the array, `_loadToHaveAtLeast(int minAvailable)` was called from `_decodeShortAsciiValue(int len)`.  In the particular case where the location appeared to go backwards, both `_inputEnd` and `_inputPtr` were set to 8000 and `minAvailable` was set to 26.

The following if block appears to not update `_currInputProcessed` if `_inputEnd` and `_inputPtr` are set to the same value (i.e., if we're at the end of the current set of buffered data):

```java
// Need to move remaining data in front?
int amount = _inputEnd - _inputPtr;
if (amount > 0 && _inputPtr > 0) {
    _currInputProcessed += _inputPtr;
    //_currInputRowStart -= _inputPtr;
    System.arraycopy(_inputBuffer, _inputPtr, _inputBuffer, 0, amount);
    _inputEnd = amount;
} else {
    _inputEnd = 0;
}
```
Since the subsequent logic reads new data into the buffer, should `_currInputProcessed` update no matter the value of `amount`?

**Comment by cowtowncoder** [comment]
Yes, I think you are right that `_inputPtr` should be added regardless.

I will add a test to verify this, and fix appropriately. Thanks!

**Comment by cowtowncoder** [comment]
There were 2 bugs, one mentioned above, and another small one that affected end-of-input handling.

Thank you for reporting this! Fix will be in 2.5.1 (and 2.4.6 if that gets released at some point -- 2.4.5 was just released, so that may take a while)

**Comment by jmcmahon356** [comment]
Thanks for the quick turnaround
